### PR TITLE
NFT Refactor: simpleHashNFTToInternalNFT

### PR DESCRIPTION
**Pull request by benisgold**
Fixes APP-####

## What changed (plus any additional context for devs)
The model we use for NFTs in Rainbow, `UniqueAsset`, is not good. Its structure is heavily based off of / relies upon opensea's nft api response, which we no longer use. Numerous attributes have missing data but are still used all around the app. `UniqueAsset` is also just generally disorganized.

A few weeks ago I introduced the `NFT`, a new internal NFT type. I did my best to keep it agnostic to the data provider (simplehash). It contains basically the same data as a `UniqueAsset`– most of the changes are structural/naming. `NFT` is not yet in use.

This PR just contains the `simpleHashNFTToInternalNFT` method, which is used to convert an NFT returned by the simplehash api (`SimpleHashNFT`) to an `NFT`.

### CODE REVIEW

File: src/handlers/web3.ts [hunk 1]
```diff
@@ -44,6 +44,9 @@ import { fetchContractABI } from '@/utils/ethereumUtils';
 import { logger, RainbowError } from '@/logger';
 import { IS_IOS } from '@/env';
 
+export const ERC1155 = 'ERC1155';
```
Comment (benisgold) [inline src/handlers/web3.ts:47]: should this be inside a `NFTStandard` enum?

File: src/resources/nfts/simplehash/utils.ts [hunk 2]
```diff
@@ -193,3 +204,206 @@ export function simpleHashNFTToUniqueAsset(nft: SimpleHashNFT): UniqueAsset {
...
+            !isNil(trait.trait_type) && !isNil(trait.value)
+        )
+        ?.map(
+          (trait: SimpleHashTrait) =>
+            ({
+              displayType: trait?.display_type,
+              traitType: trait.trait_type,
+              value: trait.value,
+            } as NFTTrait)
+        ) ?? [],
```
Comment (estrattonbailey) [inline src/resources/nfts/simplehash/utils.ts:400]: Nit: all these filter/maps, can we put those within the function body and not within the returned object? Just harder to read down here.

File: src/resources/nfts/simplehash/utils.ts [hunk 3]
```diff
@@ -193,3 +204,206 @@ export function simpleHashNFTToUniqueAsset(nft: SimpleHashNFT): UniqueAsset {
...
+            collectionId: marketplace?.marketplace_collection_id,
+            collectionUrl: marketplace?.collection_url,
+            marketplaceId: getInternalMarketplaceIdFromSimpleHashMarketplaceId(
+              marketplace.marketplace_id
+            ),
+            name: marketplace?.marketplace_name,
+            nftUrl: marketplace?.nft_url,
+          } as NFTMarketplace)
+      ),
+    name: nft.name ?? undefined,
```
Comment (estrattonbailey) [inline src/resources/nfts/simplehash/utils.ts:382]: How could an NFT not have a name?

File: src/resources/nfts/simplehash/utils.ts [hunk 4]
```diff
@@ -193,3 +204,206 @@ export function simpleHashNFTToUniqueAsset(nft: SimpleHashNFT): UniqueAsset {
...
+
+  let uniqueTokenType: UniqueTokenType = uniqueTokenTypes.NFT;
+  if (lowercasedContractAddress === POAP_NFT_ADDRESS) {
+    uniqueTokenType = uniqueTokenTypes.POAP;
+  } else if (lowercasedContractAddress === ENS_NFT_CONTRACT_ADDRESS) {
+    uniqueTokenType = uniqueTokenTypes.ENS;
+  }
+
+  return uniqueTokenType;
+}
```
Comment (estrattonbailey) [inline src/resources/nfts/simplehash/utils.ts:256]: Nit: switch might be better here, like your others

**Review by estrattonbailey — COMMENTED** [review]
Looking great!

**Review by jkadamczyk — COMMENTED** [review]
Mostly agree with Eric's comments, couldn't figure out aything else than that!

File: src/resources/nfts/types.ts [hunk 5]
```diff
@@ -8,55 +8,72 @@ export enum NFTMarketplaceId {
...
+};
+
+export type NFTMarketplace = {
+  collectionId: string | undefined;
+  collectionUrl: string | undefined;
   marketplaceId: NFTMarketplaceId;
-  name: string | null;
-  nftUrl: string | null;
+  name: string | undefined;
+  nftUrl: string | undefined;
```
[the remark this answers is not in the document]
Comment (benisgold) [reply src/resources/nfts/types.ts:23]: @estrattonbailey `filterSimpleHashNFTs` filters out all nfts without essential properties such as `name`, but `simpleHashNFTToInternalNFT` isn't aware of that. do you think it's fine to assert `name: name as string` so i can remove `undefined` from the type?

File: src/resources/nfts/simplehash/utils.ts [hunk 6]
```diff
@@ -193,3 +204,206 @@ export function simpleHashNFTToUniqueAsset(nft: SimpleHashNFT): UniqueAsset {
...
+    fm: 'png',
+    w: FULL_NFT_IMAGE_SIZE,
+  });
+  const lowResPngUrl = maybeSignUri(nonSVGUrl, {
+    fm: 'png',
+    w: CardSize,
+  });
+  const fullResUrl = isSVG && original ? original : fullResPngUrl;
+
+  return { fullResPngUrl, lowResPngUrl, fullResUrl };
```
[the remark this answers is not in the document]
Comment (benisgold) [reply src/resources/nfts/simplehash/utils.ts:255]: i don't think svgs are being signed here

File: src/resources/nfts/simplehash/utils.ts (hunk 3 above)
Comment (benisgold) [reply src/resources/nfts/simplehash/utils.ts:430]: bc a `SimpleHashNFT` defines `name: string | undefined`. I put a note in the docblock saying that this function may produce an invalid NFT if `filterSimpleHashNFTs` is not called first (that's where NFTs with undefined names are removed).

File: src/resources/nfts/simplehash/utils.ts [hunk 7]
```diff
@@ -193,3 +204,206 @@ export function simpleHashNFTToUniqueAsset(nft: SimpleHashNFT): UniqueAsset {
...
+        ?.map(
+          (trait: SimpleHashTrait) =>
+            ({
+              displayType: trait?.display_type,
+              traitType: trait.trait_type,
+              value: trait.value,
+            } as NFTTrait)
+        ) ?? [],
+    type: AssetTypes.nft as AssetType,
+    uniqueId: `${network}_${nft.contract_address}_${nft.token_id}`,
```
[the remark this answers is not in the document]
Comment (benisgold) [reply src/resources/nfts/simplehash/utils.ts:444]: https://github.com/rainbow-me/rainbow/pull/4755#discussion_r1161016806

File: src/resources/nfts/types.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (benisgold) [reply src/resources/nfts/types.ts:23]: unless you have a differing opinion i will keep `name: string | undefined` so we don't need to rely on certain `SimpleHashNFTs` being filtered out in order to ensure type correctness.

**Review by estrattonbailey — APPROVED** [review]

File: src/components/expanded-state/unique-token/UniqueTokenExpandedStateContent.js [hunk 8]
```diff
@@ -46,6 +46,7 @@ const UniqueTokenExpandedStateContent = ({
   // default to showing a loading spinner for 3D/video assets
   const [loading, setLoading] = React.useState(supports3d || supportsVideo);
 
+  console.log(supportsVideo);
```
Comment (jkadamczyk) [inline src/components/expanded-state/unique-token/UniqueTokenExpandedStateContent.js:49]: Do we need this log?

File: src/handlers/svgs.ts [hunk 9]
```diff
@@ -20,7 +20,7 @@ cloudinaryConfig({
 const pixelRatio = PixelRatio.get();
 const RAINBOW_PROXY = 'https://images.rainbow.me/proxy?url=';
 
-function svgToPng(url: any, big = false) {
+function svgToPng(url: string, big = false) {
```
Comment (jkadamczyk) [inline src/handlers/svgs.ts:23]: 👍

File: src/handlers/svgs.ts [hunk 10]
```diff
@@ -41,7 +41,11 @@ function svgToPng(url: any, big = false) {
   return cloudinaryUrl;
 }
 
-export default function svgToPngIfNeeded(url: any, big: any) {
+export default function svgToPngIfNeeded(
+  url: string | null | undefined,
+  big?: boolean
```
Comment (jkadamczyk) [inline src/handlers/svgs.ts:46]: Thanks for typing this!

**Review by jkadamczyk — APPROVED** [review]

**Pull request merged**